lib/rpa_paycode.py

```diff
@@ -0,0 +1,340 @@
+            raise ValueError("op_return_raw must be an even number of hex digits") from e
+        final_outputs.append(OPReturn.output_for_rawhex(op_return_raw))
+         
+    for address, amount in outputs:
+        address = resolver(wallet, address, nocheck)
+        amount = satoshis(amount)
+        final_outputs.append((TYPE_ADDRESS, address, amount))
+         
+    coins = wallet.get_spendable_coins(domain, config)
+    tx = wallet.make_unsigned_transaction(coins, final_outputs, config, fee, change_addr)
+    if locktime != None:
+        tx.locktime = locktime
+    if not unsigned:
+        run_hook('sign_tx', wallet, tx) 
+        wallet.sign_transaction(tx, password)
+    return tx
+
+def calculate_paycode_shared_secret(private_key, public_key, outpoint):
+    
+    # private key is expected to be an integer.
+    # public_key is expected to be bytes.
+    # outpoint is expected to be a string.
+    # returns the paycode shared secret as bytes
+        
+    from fastecdsa import keys, curve
+    from fastecdsa.point import Point 
+        
+    # Public key is expected to be compressed.  Change into a point object.   
+    pubkey_point = bitcoin.ser_to_point(public_key)
+    fastecdsa_point = Point(pubkey_point.x(), pubkey_point.y(), curve.secp256k1) 
+        
+    # Multiply the public and private points together
+    ecdh_product = fastecdsa_point * private_key
+    ecdh_x = ecdh_product.x
+    ecdh_x_bytes = ecdh_x.to_bytes(33, byteorder="big")
+        
+    # Get the hash of the product
+    sha_ecdh_x_bytes = sha256(ecdh_x_bytes)
+    sha_ecdh_x_as_int = int.from_bytes(sha_ecdh_x_bytes, byteorder="big")
+  
+    # Hash the outpoint string
+    hash_of_outpoint = sha256(outpoint)
+    hash_of_outpoint_as_int = int.from_bytes(hash_of_outpoint, byteorder="big")
+        
+    # Sum the ECDH hash and the outpoint Hash 
+    grand_sum = sha_ecdh_x_as_int + hash_of_outpoint_as_int
+
+    # Hash the final result
+    grand_sum_hex = hex(grand_sum)
+    shared_secret = sha256(grand_sum_hex) 
+        
+    return shared_secret
+
+def generate_address_from_pubkey_and_secret(parent_pubkey, secret):
+  
+    # parent_pubkey and secret are expected to be bytes
+    # This function generates a receiving address based on CKD.
+        
+    new_pubkey = bitcoin.CKD_pub(parent_pubkey, secret, 0)[0]
+        
+    use_uncompressed = False
+
+    # Currently, just uses compressed keys, but if this ever changes to require uncompressed points:
+    if use_uncompressed:
+        pubkey_point = bitcoin.ser_to_point(new_pubkey)
+        point_x=pubkey_point.x()
+        point_y=pubkey_point.y()
+        uncompressed="04"+hex(pubkey_point.x())[2:]+hex(pubkey_point.y())[2:] 
+        new_pubkey = bytes.fromhex(uncompressed)
+                    
+    addr = Address.from_pubkey(new_pubkey)
+    return addr
+        
+         
+def generate_privkey_from_secret(parent_privkey, secret):
+  
+    # parent_privkey and secret are expected to be bytes
```

Comment:
It's customary to do python method docs as so:

```python
def generate_privkey_from_secret(parent_privkey, secret):
    """parent_privkey and secret are expected to be bytes.
       This function generates a receiving address based on CKD."""

Reply:
done in 060ca8f